### PR TITLE
Scenario lifecycle hooks update

### PR DESCRIPTION
**Pull request by sleal-unity**
# Peer Review Information:

**NOTICE**: I found two commits I made 4 days ago using a different GitHub user name that I've amended in this PR. Everything should merge properly, but you'll see an interestingly long commit history being listed here, when in reality, there's just one commit actually being made.

This PR refactors the existing scenario lifecycle hooks to make custom scenarios easier to derive and prepare the scenario class to enable more ADR-like data streaming workflows. This PR also adds the PerceptionScenario abstract class to better distinguish the scenario's coordinator responsibilities from basic perception functionality.

## Editor / Package versioning:
**Editor Version Target**: 2019.4

## Dev Testing:
**Tests Added**: 
Existing tests have been updated to use the new lifecycle hooks

## Checklist
- [X] - Updated docs
- [X] - Updated changelog

### CODE REVIEW

File: com.unity.perception/Runtime/Randomization/Scenarios/PerceptionScenario.cs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        protected override void OnComplete()
+        {
+            DatasetCapture.ResetSimulation();
+            Manager.Instance.Shutdown();
+        }
+
+        /// <inheritdoc/>
+        protected override void OnIdle()
+        {
+            if (Manager.FinalUploadsDone)
```
Comment (JonathanHUnity) [inline com.unity.perception/Runtime/Randomization/Scenarios/PerceptionScenario.cs:81]: This should not be necessary. Calling Quit() will automatically call Manager.Instance.Shutdown() and should terminate synchronously.

File: com.unity.perception/Runtime/Randomization/Scenarios/PerceptionScenario.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                Quit();
+        }
+
+        /// <summary>
+        /// Used to report a scenario iteration as a perception metric
+        /// </summary>
+        protected struct IterationMetricData
```
Comment (JonathanHUnity) [inline com.unity.perception/Runtime/Randomization/Scenarios/PerceptionScenario.cs:88]: why protected? Nested types in the public API should be avoided.

File: com.unity.perception/Runtime/Randomization/Scenarios/PerceptionScenario.cs (hunk 1 above)
Comment (sleal-unity) [reply com.unity.perception/Runtime/Randomization/Scenarios/PerceptionScenario.cs:81]: Ahh, I'll need to test this again then. In the past, Devin and I ran into a problem with how the final uploads took a few frames to complete before we could quit, or else the simulation would not save all the recorded data properly.

File: com.unity.perception/Runtime/Randomization/Scenarios/PerceptionScenario.cs (hunk 2 above)
Comment (sleal-unity) [reply com.unity.perception/Runtime/Randomization/Scenarios/PerceptionScenario.cs:88]: My thought was that someone deriving this class would maybe want to either not record this metric or potentially rearrange the timing a bit, and therefore require the class to be accessible in some way so they could use it in their derived class.

Comment (sleal-unity) [reply com.unity.perception/Runtime/Randomization/Scenarios/PerceptionScenario.cs:88]: What do you think about just moving it outside of the PerceptionScenarioClass and making it public?

File: com.unity.perception/Runtime/Randomization/Scenarios/PerceptionScenario.cs (hunk 1 above)
Comment (sleal-unity) [reply com.unity.perception/Runtime/Randomization/Scenarios/PerceptionScenario.cs]: Yep, this totally works. The scenario now quits within the OnComplete() hook

File: com.unity.perception/Runtime/Randomization/Scenarios/PerceptionScenario.cs (hunk 2 above)
Comment (sleal-unity) [reply com.unity.perception/Runtime/Randomization/Scenarios/PerceptionScenario.cs]: I just made it private. It's not a big deal.

**Pull request closed without merging**